File: Python/algorithmic_toolbox/14_distinct_decomposition.py
import sys


def get_decomposition_to_maximal_number_of_distinct_elems(n):
    summands = []
    smallest_summand = 1
    while n > 0:
        if n > 2 * smallest_summand:
            elem = smallest_summand
        else:
            elem = n
        summands.append(elem)
        n -= elem
        smallest_summand += 1
    return summands


def main():
    data = sys.stdin.read().splitlines()
    n = int(data[0])
    result = get_decomposition_to_maximal_number_of_distinct_elems(n)
    print(len(result))
    print(" ".join([str(elem) for elem in result]))


if __name__ == "__main__":
    main()

File: R/algorithmic_toolbox/14_distinct_decomposition.R
#!/usr/bin/env Rscript
DecomposeToDistinctElems <- function(n) {
    smallestSummand <- 1
    summands <- vector(mode = "integer", length = as.integer(sqrt(2*n)))
    while (n > 0) {
        elem <- if (n > 2 * smallestSummand) smallestSummand else n
        summands[smallestSummand] <- elem
        n <- n - elem
        smallestSummand <- smallestSummand + 1
    }
    summands[summands > 0]
}

Main <- function() {
    conn <- file("stdin", "r")
    n <- as.integer(readLines(conn, n = 1))
    close(conn)
    result <- DecomposeToDistinctElems(n)
    cat(length(result), fill = TRUE)
    cat(result, fill = TRUE)
}

if (!interactive()) {
    Main()
}
